# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources (README.md file)
- Happy learning!

In [1]:
# import numpy and pandas
import pandas as pd
import numpy as np
from scipy.stats import trim_mean, mode, skew, gaussian_kde, pearsonr, spearmanr, beta
from statsmodels.stats.weightstats import ztest as ztest

from scipy.stats import ttest_ind, norm, t
from scipy.stats import f_oneway
from scipy.stats import sem

# Challenge 1 - Exploring the Data

In this challenge, we will examine all salaries of employees of the City of Chicago. We will start by loading the dataset and examining its contents

In [3]:
# Run this code:
salaries = pd.read_csv(r"C:\Users\nahia\OneDrive\Escritorio\Ironhack_Bootcamp\Woche_Vier\lab-hypothesis-testing-en\data\Current_Employee_Names__Salaries__and_Position_Titles.csv")

Examine the `salaries` dataset using the `head` function below.

In [4]:
# Your code here
salaries.head()

Name                              Job Titles  \
0    AARON,  JEFFERY M                                SERGEANT   
1      AARON,  KARINA   POLICE OFFICER (ASSIGNED AS DETECTIVE)   
2  AARON,  KIMBERLEI R                CHIEF CONTRACT EXPEDITER   
3  ABAD JR,  VICENTE M                       CIVIL ENGINEER IV   
4    ABASCAL,  REECE E             TRAFFIC CONTROL AIDE-HOURLY   

         Department Full or Part-Time Salary or Hourly  Typical Hours  \
0            POLICE                 F           Salary            NaN   
1            POLICE                 F           Salary            NaN   
2  GENERAL SERVICES                 F           Salary            NaN   
3       WATER MGMNT                 F           Salary            NaN   
4              OEMC                 P           Hourly           20.0   

   Annual Salary  Hourly Rate  
0       101442.0          NaN  
1        94122.0          NaN  
2       101592.0          NaN  
3       110064.0          NaN  
4            NaN        19.86

# Challenge 2 - Hypothesis Tests

In this section of the lab, we will test whether the hourly wage of all hourly workers is significantly different from $30/hr. Import the correct one sample test function from scipy and perform the hypothesis test for a 95% two sided confidence interval.

In [5]:
# Your code here
!pip install --upgrade scipy

   ---------------------------------------- 0.0/38.5 MB ? eta -:--:--
   ---------------------------------------- 0.3/38.5 MB ? eta -:--:--
    --------------------------------------- 0.8/38.5 MB 2.5 MB/s eta 0:00:16
   - -------------------------------------- 1.3/38.5 MB 2.6 MB/s eta 0:00:15
   -- ------------------------------------- 2.1/38.5 MB 2.8 MB/s eta 0:00:14
   -- ------------------------------------- 2.6/38.5 MB 2.8 MB/s eta 0:00:13
   --- ------------------------------------ 3.4/38.5 MB 3.0 MB/s eta 0:00:12
   ---- ----------------------------------- 4.2/38.5 MB 3.1 MB/s eta 0:00:12
   ----- ---------------------------------- 5.2/38.5 MB 3.3 MB/s eta 0:00:11
   ------ --------------------------------- 6.0/38.5 MB 3.4 MB/s eta 0:00:10
   ------- -------------------------------- 7.1/38.5 MB 3.6 MB/s eta 0:00:09
   -------- ------------------------------- 8.1/38.5 MB 3.7 MB/s eta 0:00:09
   --------- ------------------------------ 9.4/38.5 MB 3.9 MB/s eta 0:00:08
   ----------


[notice] A new release of pip is available: 25.2 -> 25.3
[notice] To update, run: C:\Users\nahia\anaconda3\python.exe -m pip install --upgrade pip


In [22]:
import pandas as pd
import numpy as np
import scipy.stats as st
import matplotlib.pyplot as plt
import seaborn as sns

In [23]:
salaries = pd.read_csv(r"C:\Users\nahia\OneDrive\Escritorio\Ironhack_Bootcamp\Woche_Vier\lab-hypothesis-testing-en\data\Current_Employee_Names__Salaries__and_Position_Titles.csv")
salaries.head()
salaries.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 33183 entries, 0 to 33182
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Name               33183 non-null  object 
 1   Job Titles         33183 non-null  object 
 2   Department         33183 non-null  object 
 3   Full or Part-Time  33183 non-null  object 
 4   Salary or Hourly   33183 non-null  object 
 5   Typical Hours      8022 non-null   float64
 6   Annual Salary      25161 non-null  float64
 7   Hourly Rate        8022 non-null   float64
dtypes: float64(3), object(5)
memory usage: 2.0+ MB


In [24]:
salaries = salaries.dropna(subset=['Hourly Rate'])

In [25]:
c1_sample = salaries['Hourly Rate'].sample(30)
st.ttest_1samp(c1_sample,30)

TtestResult(statistic=np.float64(1.4154314415171325), pvalue=np.float64(0.167592128384605), df=np.int64(29))

In [26]:
0.167592128384605<0.05

False

There is no evidence that the average wage differs from $30/hr.

# Challenge 3 - Constructing Confidence Intervals

While testing our hypothesis is a great way to gather empirical evidence for accepting or rejecting the hypothesis, another way to gather evidence is by creating a confidence interval. A confidence interval gives us information about the true mean of the population. So for a 95% confidence interval, we are 95% sure that the mean of the population is within the confidence interval. 
).

To read more about confidence intervals, click [here](https://en.wikipedia.org/wiki/Confidence_interval).


In the cell below, we will construct a 95% confidence interval for the mean hourly wage of all hourly workers. 

The confidence interval is computed in SciPy using the `t.interval` function. You can read more about this function [here](https://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.stats.t.html).

To compute the confidence interval of the hourly wage, use the 0.95 for the confidence level, number of rows - 1 for degrees of freedom, the mean of the sample for the location parameter and the standard error for the scale. The standard error can be computed using [this](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.sem.html) function in SciPy.

In [34]:
import math
from scipy import stats

In [28]:
salaries = salaries.dropna(subset=['Hourly Rate'])

In [50]:
# Your code here
sigma = salaries['Hourly Rate'].std()/math.sqrt(30)

stats.t.interval( 0.95,                                     #confidence level
                  df = 29,                                   #Degrees of freedom
                  loc = salaries['Hourly Rate'].mean(),     #Sample mean
                  scale = sigma)                            #Standard dev estimate

(np.float64(28.26564874404275), np.float64(37.311466688517726))

Lower limit 28.26564874404275  and Upper limit 37.311466688517726

# Challenge 4 - Hypothesis Tests of Proportions

Another type of one sample test is a hypothesis test of proportions. In this test, we examine whether the proportion of a group in our sample is significantly different than a fraction. 

You can read more about one sample proportion tests [here](http://sphweb.bumc.bu.edu/otlt/MPH-Modules/BS/SAS/SAS6-CategoricalData/SAS6-CategoricalData2.html).

In the cell below, use the `proportions_ztest` function from `statsmodels` to perform a hypothesis test that will determine whether the number of hourly workers in the City of Chicago is significantly different from 25% at the 95% confidence level.

In [ ]:
import pandas as pd
from statsmodels.stats.proportion import proportions_ztest

In [66]:
salaries = pd.read_csv(r"C:\Users\nahia\OneDrive\Escritorio\Ironhack_Bootcamp\Woche_Vier\lab-hypothesis-testing-en\data\Current_Employee_Names__Salaries__and_Position_Titles.csv")

In [67]:
# Your code here

salaries.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 33183 entries, 0 to 33182
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Name               33183 non-null  object 
 1   Job Titles         33183 non-null  object 
 2   Department         33183 non-null  object 
 3   Full or Part-Time  33183 non-null  object 
 4   Salary or Hourly   33183 non-null  object 
 5   Typical Hours      8022 non-null   float64
 6   Annual Salary      25161 non-null  float64
 7   Hourly Rate        8022 non-null   float64
dtypes: float64(3), object(5)
memory usage: 2.0+ MB


In [68]:
salaries['Salary or Hourly'].value_counts()

Salary or Hourly
Salary    25161
Hourly     8022
Name: count, dtype: int64

In [70]:
obs = 33183

In [71]:
hourly = 8022

In [72]:
p0 = 0.25

In [73]:
stat, p_value = proportions_ztest(count=hourly, nobs=hourly, value=p0)

c:\Users\nahia\AppData\Local\Programs\Python\Python313\Lib\site-packages\statsmodels\stats\weightstats.py:792: RuntimeWarning: divide by zero encountered in scalar divide
  zstat = value / std


In [74]:
print(f' z: {stat:.4f}')
print(f' p value: {p_value:.4f}')

 z: inf
 p value: 0.0000


In [76]:
alpha = 0.05
if p_value < alpha:
    print(" There is significant evidence that the proportion differs from 0.25.")
else:
    print(" There is not enough evidence to conclude that the proportion differs from 0.25")

 There is significant evidence that the proportion differs from 0.25.
